embassy-usb-driver/src/host.rs

```diff
@@ -0,0 +1,309 @@
+    impl Type for Control {
+        fn ep_type() -> EndpointType {
+            EndpointType::Control
+        }
+    }
+    impl Type for Interrupt {
+        fn ep_type() -> EndpointType {
+            EndpointType::Interrupt
+        }
+    }
+    impl Type for Bulk {
+        fn ep_type() -> EndpointType {
+            EndpointType::Bulk
+        }
+    }
+    impl Type for Isochronous {
+        fn ep_type() -> EndpointType {
+            EndpointType::Isochronous
+        }
+    }
+
+    #[diagnostic::on_unimplemented(message = "This is not a CONTROL channel")]
+    pub trait IsControl {}
+    impl IsControl for Control {}
+
+    #[diagnostic::on_unimplemented(message = "This is not a CONTROL channel")]
+    pub trait IsInterrupt {}
+    impl IsInterrupt for Interrupt {}
+
+    pub trait Direction {
+        fn is_in() -> bool;
+        fn is_out() -> bool;
+    }
+    pub struct In {}
+    pub struct Out {}
+    pub struct InOut {}
+    impl Direction for In {
+        fn is_in() -> bool {
+            true
+        }
+        fn is_out() -> bool {
+            false
+        }
+    }
+    impl Direction for Out {
+        fn is_in() -> bool {
+            false
+        }
+        fn is_out() -> bool {
+            true
+        }
+    }
+    impl Direction for InOut {
+        fn is_in() -> bool {
+            true
+        }
+        fn is_out() -> bool {
+            true
+        }
+    }
+
+    #[diagnostic::on_unimplemented(message = "This is not an IN channel")]
+    pub trait IsIn: Direction {}
+    impl IsIn for In {}
+    impl IsIn for InOut {}
+
+    #[diagnostic::on_unimplemented(message = "This is not an OUT channel")]
+    pub trait IsOut: Direction {}
+    impl IsOut for Out {}
+    impl IsOut for InOut {}
+}
+
+/// Specify the timeout of a channel
+pub struct TimeoutConfig {
+    /// Maximum response timeout for transactions with a Data Stage
+    pub data_timeout: Duration,
+    /// Maximum response timeout for transactions without a data stage
+    pub standard_timeout: Duration,
+}
+
+impl Default for TimeoutConfig {
+    fn default() -> Self {
+        TimeoutConfig {
+            data_timeout: Duration::from_millis(500),
+            standard_timeout: Duration::from_millis(50),
+        }
+    }
+}
+
+/// ## Virtual USB Channels
+/// These contain the required information to send a packet correctly to a device endpoint.
+/// The information is carried with the channel on creation (see [`UsbHostDriver::alloc_channel`]) and can be changed with [`UsbChannel::retarget_channel`].
+///
+/// It is up to the hal's driver how to implement concurrent requests, some hardware IP may allow for multiple hardware channels
+///  while others may only have a single channel which needs to be multiplexed in software, while others still use DMA request linked-lists.
+/// Any of these are compatibile with the UsbChannel with varying degrees of sync primitives required.
+pub trait UsbChannel<T: channel::Type, D: channel::Direction> {
```

Comment:
Are the generics really needed? These could be replaced with runtime checks or more traits.
```Rust
trait UsbChannel;
trait UsbChannelIn;
trait UsbChannelOut;
trait UsbChannelControl;
```